### ISSUE TITLE
Multi-index search for Knn

### ISSUE THREAD

**Issue by aluriga** [opendistro-for-elasticsearch/k-NN#340]
Hi, I would like to create two indexes with the following schema.

Index 1
ID, feature_vector1
Index 2
ID feature_vector2

Context for creating two indexes: For each ID, there can be 3-10 feature_vector1's; whereas feature_vector2 is unique.
 I want search across the both the indexes using the two feature vectors and get a combined set of knn results(which satisfy both the conditions). Is there a way to do this?

**Comment by VijayanB** [comment]
@aluriga you can perform multi index search ( provided identical mapping ) with identical search query against index pattern.

example:
Step 1. Create index1
```
curl -X PUT "localhost:9200/index1" -H 'Content-Type: application/json' -d'
{
"settings" : {
  "number_of_shards" :   1,
  "number_of_replicas" : 0,
  "index.knn" : true
  },
  "mappings": {
    "properties": {
      "feature_vector1": {
        "type": "knn_vector",
        "dimension": 4
      },
      "docID": {
        "type": "keyword"
      }
    }
  }
}
'
```
Step 2. Create index2
```
curl -X PUT "localhost:9200/index2" -H 'Content-Type: application/json' -d'
{
"settings" : {
  "number_of_shards" :   1,
  "number_of_replicas" : 0,
  "index.knn" : true
  },
  "mappings": {
    "properties": {
      "feature_vector1": {
        "type": "knn_vector",
        "dimension": 4
      },
      "docID": {
        "type": "keyword"
      }
    }
  }
}
'
```

Step 3. Insert some documents
```
curl -X PUT "localhost:9200/index1/_doc/1?pretty" -H 'Content-Type: application/json' -d'
{
  "feature_vector1": [1, 1, 0, 2],
  "docID" : "doc1"
}
'
curl -X PUT "localhost:9200/index1/_doc/2?pretty" -H 'Content-Type: application/json' -d'
{
  "feature_vector1": [2, 2, 3, 4],
  "docID" : "doc2"
}
'

curl -X PUT "localhost:9200/index1/_doc/3?pretty" -H 'Content-Type: application/json' -d'
{
  "feature_vector1": [1, 1, 0, 2],
  "docID" : "doc2"
}
'
curl -X PUT "localhost:9200/index2/_doc/2?pretty" -H 'Content-Type: application/json' -d'
{
  "feature_vector1": [1, 1, 0, 3],
   "docID" : "doc8"
}'

curl -X PUT "localhost:9200/index2/_doc/1?pretty" -H 'Content-Type: application/json' -d'
{
  "feature_vector1": [1, 1, 0, 2],
   "docID" : "doc7"
}
'
```

Step 4. Search:
```
curl -X GET "localhost:9200/index*/_search?pretty"  -H 'Content-Type: application/json' -d'
{
  "size": 4,
  "query": {
    "knn": {
      "feature_vector1": {
        "vector": [1, 1, 0, 2],
        "k": 3
      }
    }
  }
}'
```

Output:
```
     {
  "took" : 3,
  "timed_out" : false,
  "_shards" : {
    "total" : 2,
    "successful" : 2,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 5,
      "relation" : "eq"
    },
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "index1",
        "_type" : "_doc",
        "_id" : "1",
        "_score" : 1.0,
        "_source" : {
          "feature_vector1" : [
            1,
            1,
            0,
            2
          ],
          "docID" : "doc1"
        }
      },
      {
        "_index" : "index1",
        "_type" : "_doc",
        "_id" : "3",
        "_score" : 1.0,
        "_source" : {
          "feature_vector1" : [
            1,
            1,
            0,
            2
          ],
          "docID" : "doc2"
        }
      },
      {
        "_index" : "index2",
        "_type" : "_doc",
        "_id" : "1",
        "_score" : 1.0,
        "_source" : {
          "feature_vector1" : [
            1,
            1,
            0,
            2
          ],
          "docID" : "doc7"
        }
      },
      {
        "_index" : "index2",
        "_type" : "_doc",
        "_id" : "2",
        "_score" : 0.5,
        "_source" : {
          "feature_vector1" : [
            1,
            1,
            0,
            3
          ],
          "docID" : "doc8"
        }
      }
    ]
  }
}
 
```
Here, the top k results are calculated across indices and sorted and returned top "size" results. Since, the scores are calculated independent of indices, i believe this would give you top 4 results across indices based on space type you used. 

P.S: In order to compare across indices, it is important the mapping is identical across all indices.

@aluriga is this what were you looking for?

**Comment by aluriga** [comment]
@VijayanB Thank you for your detailed response! My current scenario is this. 
I have an ID with a 300-dimension feature_vector1. The same ID has sub parts with another 100-dimensional feature_vector2. 

ID, SID1, feature_vector1a, feature_vector2a
ID, SID2,  feature_vector1a, feature_vector2b
ID, SID3,  feature_vector1a, feature_vector2c
ID, SID4,  feature_vector1a, feature_vector2d
ID, SID5,  feature_vector1a, feature_vector2e

Since the dataset I am dealing with is extremely large, I do not want store duplicates of the data. What I want to get is - top k results for a query feature_vector1 **and** feature_vector2. Is it possible to get this? What is the best possible way to do this?

Basically for every 1 feature_vector1, there are 5 feature_vector2's. For this reason, I thought that maybe indexes can be created separately. But I wasnt sure on how to get the combined results of the knn.

**Comment by VijayanB** [comment]
@aluriga Thank you for clarification.
Every field is stored separately for approximate knn search. You can create two sub query and have bool on top of those sub query.
Your first sub query will search top k on feature_vector1 and second query will search on feature_vector2 . The output of bool makes sure only common documents are retrieved.

```
{
  "size": 4,
  "query": {
    "bool": {
      "must": [
        {
          "knn": {
            "feature_vector1": {
              "vector": [1,1,0,2],
              "k": 3
            }
          }
        },
        {
          "knn": {
            "feature_vector2": {
              "vector": [ 1,1],
              "k": 3
            }
          }
        }
      ]
    }
  }
}
```
I assume that feature2a, feature2b can be found across feature1a and feature 1b, thats why you would like to have both to be part of the search query instead of feature2 field alone.

To reduce storage, have you explored nested field type? You can create feature2 as nested field inside feature1 ( provided every other fields are common ).
Will this solve your concerns?

**Comment by aluriga** [comment]
@VijayanB Thank you for the reply. 

> Every field is stored separately for approximate knn search. You can create two sub query and have bool on top of those sub query.

Does this mean in the same index I can put feature_vector1 and feature_vector2 at different times for the same ID/doc?

> To reduce storage, have you explored nested field type? You can create feature2 as nested field inside feature1 ( provided every other fields are common ).

Can you please provide an example on how to do this?

**Comment by VijayanB** [comment]
knn_vector is like any other field type. You can insert document by adding feature_vector1 and later, update same doc by adding feature_vector2.

Example for nested:
We will create an index with following mapping
docID,feature_vector1a, feature { [feature_vector2a, feature_vector2b,...]}

```
curl -X PUT "localhost:9200/index2" -H 'Content-Type: application/json' -d'
{
  "settings": {
    "number_of_shards": 1,
    "number_of_replicas": 0,
    "index.knn": true
  },
  "mappings": {
    "properties": {
      "feature_vector1": {
        "type": "knn_vector",
        "dimension": 4
      },
      "feature": {
        "type": "nested",
        "properties": {
          "vector2": {
            "type": "knn_vector",
            "dimension": 4
          }
        }
      },
      "docID": {
        "type": "keyword"
      }
    }
  }
}'
```

Insert documents

```
curl -X PUT "localhost:9200/index2/_doc/1?pretty" -H 'Content-Type: application/json' -d'
{
  "feature_vector1": [1, 1, 0, 2],
  "feature": [
  {
    "vector2": [1, 2, 3, 4]
  },
  {
    "vector2": [1, 2, 3, 5]
  },
  {
    "vector2": [1, 2, 3, 6]
  },
  {
    "vector2": [1, 2, 3, 7]
  },
  {
    "vector2": [1, 2, 3, 8]
  }
 ],
  "docID" : "doc1"
}
'

curl -X PUT "localhost:9200/index2/_doc/2?pretty" -H 'Content-Type: application/json' -d'
{
  "feature_vector1": [1, 1, 0, -1],
  "feature": [
  {
    "vector2": [1, 2, 3, 4]
  },
  {
    "vector2": [1, 2, 3, 5]
  },
  {
    "vector2": [1, 2, 3, 6]
  },
  {
    "vector2": [1, 2, 3, 7]
  },
  {
    "vector2": [1, 2, 3, 8]
  }
 ],
  "docID" : "doc1"
}
'
```

search document with feature_vector1 & feature_vector2
```
curl -X GET "localhost:9200/index2/_search?pretty"  -H 'Content-Type: application/json' -d'
{
  "size": 4,
  "query": {
    "bool": {
      "must": [
        {
          "knn": {
            "feature_vector1": {
              "vector": [1,1,0,2],
              "k": 2
            }
          }
        },
        {
          "nested": {
            "path": "feature",
            "query": {
              "knn": {
                "feature.vector2": {
                  "vector": [1,2,3,4],
                  "k": 2
                }
              }
            }
          }
        }
      ]
    }
  }
}'
```
The above search result will sum scores based on match from both the query and sort the results.

**Comment by VijayanB** [comment]
@aluriga let us know if that helps.

**Comment by aluriga** [comment]
@VijayanB Thank you for the detailed examples. This is very helpful. I will try these ideas for my use case and get back to you if I have any problems or difficulties.

**Comment by aluriga** [comment]
> 
> Insert documents
> 
> ```
> curl -X PUT "localhost:9200/index2/_doc/1?pretty" -H 'Content-Type: application/json' -d'
> {
>   "feature_vector1": [1, 1, 0, 2],
>   "feature": [
>   {
>     "vector2": [1, 2, 3, 4]
>   },
>   {
>     "vector2": [1, 2, 3, 5]
>   },
>   {
>     "vector2": [1, 2, 3, 6]
>   },
>   {
>     "vector2": [1, 2, 3, 7]
>   },
>   {
>     "vector2": [1, 2, 3, 8]
>   }
>  ],
>   "docID" : "doc1"
> }
> '
> 
> curl -X PUT "localhost:9200/index2/_doc/2?pretty" -H 'Content-Type: application/json' -d'
> {
>   "feature_vector1": [1, 1, 0, -1],
>   "feature": [
>   {
>     "vector2": [1, 2, 3, 4]
>   },
>   {
>     "vector2": [1, 2, 3, 5]
>   },
>   {
>     "vector2": [1, 2, 3, 6]
>   },
>   {
>     "vector2": [1, 2, 3, 7]
>   },
>   {
>     "vector2": [1, 2, 3, 8]
>   }
>  ],
>   "docID" : "doc1"
> }
> '
> ```

@VijayanB While trying to the insert documents in this fashion, I somehow am not able to get this to work. I can see that it saves only the last entry. This is what I have been suspecting. Since the key("vector2") is same, how can I store 5 values of the same ID each with a corresponding sub-ID?
If it was just 1 vector1 and 1 vector2 for an ID, the nested schema, query all seem to work as expected.

**Comment by VijayanB** [comment]
@aluriga i am not sure why you were not able to update. Can you post your error message? Also when i try to update, i can add new element to existing array using update api.
For example, after i inserted both documents, lets say i want to add another feature [1,2,3,-1] to _doc=1.
```
curl -X POST "localhost:9200/index2/_doc/1/_update?pretty" -H 'Content-Type: application/json' -d'
{
  "script": {
    "source": "ctx._source.feature.add(params.data)",
    "lang": "painless",
    "params": {
      "data": {
        "vector2": [1,2,3,-1]
      }
    }
  }
}'
```

output:
```
{
  "_index" : "index2",
  "_type" : "_doc",
  "_id" : "1",
  "_version" : 2,
  "_seq_no" : 1,
  "_primary_term" : 1,
  "found" : true,
  "_source" : {
    "feature_vector1" : [
      1,
      1,
      0,
      2
    ],
    "feature" : [
      {
        "vector2" : [
          1,
          2,
          3,
          4
        ]
      },
      {
        "vector2" : [
          1,
          2,
          3,
          5
        ]
      },
      {
        "vector2" : [
          1,
          2,
          3,
          6
        ]
      },
      {
        "vector2" : [
          1,
          2,
          3,
          7
        ]
      },
      {
        "vector2" : [
          1,
          2,
          3,
          8
        ]
      },
      {
        "vector2" : [
          1,
          2,
          3,
          -1
        ]
      }
    ],
    "docID" : "doc1"
  }
}
```

**Comment by aluriga** [comment]
@VijayanB This is what I have.
```
knn_index = {
        'settings': {
            'index':{
                'knn': True,
                'knn.space_type': 'cosinesimil'
            },
            'translog.flush_threshold_size': '10gb',
            'refresh_interval': '-1',
            'number_of_replicas': 0
        },
        'mappings': {
            'properties': {
                'feature_vector1': {
                    'type': 'knn_vector',
                    'dimension': 3
                },
                'features':{
                    'type':'nested',
                    'properties':{
                        'feature_vector2':{
                            'type':'knn_vector',
                            'dimension':4
                        }
                    }
                }
            }
        }
    }
es.indices.create(index=index_name, body=knn_index, ignore=400)

doc = {'feature_vector1': [1,-1,2],
       'features': [
          {'feature_vector2':[1,1,0,4],
          'feature_vector2':[1,3,4,2]}
      ]}
es.index(index='test-v1',id='ID1', body=doc)
```
I see that the document is inserted successfully with the message below.
```
**Output:**
{'_index': 'test-v1',
 '_type': '_doc',
 '_id': 'ID1',
 '_version': 1,
 'result': 'created',
 '_shards': {'total': 1, 'successful': 1, 'failed': 0},
 '_seq_no': 0,
 '_primary_term': 1}
```
When I try to search for records randomly, to see what is inserted with the query below
```
search_random = {
    "query": {
        "function_score": {
            "random_score": {
                "seed": 10
            }
        }
    }
}


result = es.search(index='test-v1', body=search_random)
print(result)
```
This is the output I see. Only one 
```
**Output:**
{'took': 17, 'timed_out': False, '_shards': {'total': 5, 'successful': 5, 'skipped': 0, 'failed': 0}, 'hits': {'total': {'value': 1, 'relation': 'eq'}, 'max_score': 0.8543731, 'hits': [{'_index': 'test-v1', '_type': '_doc', '_id': 'ID1', '_score': 0.8543731, '_source': {'feature_vector': [1, -1, 2], 'features': [{'feature_vector2': [1, 3, 4, 2]}]}}]}}
```

**Comment by VijayanB** [comment]
@aluriga doc should be like below
```
doc = {'feature_vector1': [1,-1,2],
       'features': [
          {
               'feature_vector2':[1,1,0,4]
          },
          {
             'feature_vector2':[1,3,4,2]
           }
      ]}
```
see feature_vector2 is inside its own {}

**Comment by aluriga** [comment]
@VijayanB Thank you! I was able to make it work. 
I have a few other questions.
1. If I want to modify the query to get 1000 neighbors based on vector1, and score these neighbors based on vector2. How can I do that? 
2. Is there a way add weight or importance to the vectors?

```
curl -X GET "localhost:9200/index2/_search?pretty"  -H 'Content-Type: application/json' -d'
{
  "size": 4,
  "query": {
    "bool": {
      "must": [
        {
          "knn": {
            "feature_vector1": {
              "vector": [1,1,0,2],
              "k": 2
            }
          }
        },
        {
          "nested": {
            "path": "feature",
            "query": {
              "knn": {
                "feature.vector2": {
                  "vector": [1,2,3,4],
                  "k": 2
                }
              }
            }
          }
        }
      ]
    }
  }
}'

```

**Comment by VijayanB** [comment]
@aluriga You are welcome. You can use [boost](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-boosting-query.html) to increase the weight (score ) of an query. To get 1000 neighbors, you can set k = 1000 (plugin supports a maximum k value of 10,000) for feature_vector1.  I would recommend explore [function score](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-function-score-query.html) . You can create multiple query and boost based on your logic. I hope this helps.
